### ISSUE TITLE
Class decorator misclassified as untyped

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#5188]
**Describe the bug**
```python
from collections.abc import Callable
from typing import TypeVar

T = TypeVar("T")


def configurable() -> Callable[[type[T]], type[T]]:
    ...


@configurable() # False positive: Untyped class decorator obscures type of class; ignoring decorator
class Foo:
    ...
```

**Expected behavior**
No error. The behavior started happening with latest pyright version 1.1.310. 1.1.309 type checks this with no issue.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

The issue is that you're using a TypeVar `T` that is bound to the function `configurable`. When you call `configurable()`, this bound TypeVar remains unsolved (because it isn't referenced by any input parameters). Pyright has some special logic that rebinds unsolved type variables to returned callables in specific cases. It was a little "loose" with this logic (which meant that it was applying this special-case logic in cases where it shouldn't). I tightened up this logic in 1.1.310, which is where the regression was introduced. It broke in cases where you're using `type[T]` rather than `T`, as you do in your code sample. I've fixed the issue and added a test case to cover this.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.311, which I just published. It will also be included in a future release of pylance.
